### PR TITLE
Mltpl asset type

### LINKED ISSUE #116: Allow Campaign Owners to Define Donation Asset Type
Currently, all campaigns use a single global donation token defined in the contract state. This feature request adds the ability for campaign owners to specify which asset/token they want to collect donations in when creating their campaign.
Current Behavior

All campaigns use the same donation token (self.donation_token.read())
Campaign owners cannot choose their preferred asset for donations
The donation token is set globally at the contract level

Proposed Enhancement

Allow campaign owners to specify their preferred donation asset when creating a campaign through the create_campaign function.
Implementation Details
Function Signature Update

```
fn _create_campaign(
    ref self: ContractState,
    campaign_ref: felt252,
    target_amount: u256,
    donation_token: ContractAddress,  // NEW PARAMETER
) -> u256

```
## Changes Required

 Update create_campaign function signature to accept donation_token parameter
 Add validation for the donation token address

Ensure the token address is not zero
Optionally validate that the token contract exists

 Modify campaign creation logic to use the provided token instead of global default
 Update existing validation to ensure the new parameter is properly handled
 Maintain backward compatibility considerations

Current Code Structure

```
let campaign = Campaigns {
    campaign_id,
    owner: caller,
    target_amount,
    current_balance,
    withdrawn_amount,
    campaign_reference: campaign_ref,
    is_closed: false,
    is_goal_reached: false,
    donation_token: self.donation_token.read(), // CHANGE THIS LINE
    is_cancelled: false,
};
```

Proposed Code Structure

```
let campaign = Campaigns {
    campaign_id,
    owner: caller,
    target_amount,
    current_balance,
    withdrawn_amount,
    campaign_reference: campaign_ref,
    is_closed: false,
    is_goal_reached: false,
    donation_token: donation_token, // USE PROVIDED TOKEN
    is_cancelled: false,
};
```

## Additional Considerations

 Validation Functions: Add helper functions to validate token addresses
 Error Handling: Define appropriate error messages for invalid token addresses
 Documentation: Update function documentation to reflect new parameter
 Testing: Comprehensive tests for different token types and edge cases
 Migration: Consider how existing campaigns will be affected

## Acceptance Criteria

 Campaign owners can specify any valid ERC20 token address for donations
 Invalid token addresses are properly rejected with clear error messages
 Existing validation logic (campaign_ref, target_amount) remains intact
 All existing tests continue to pass
 New tests cover the donation token validation scenarios
 Function maintains the same return behavior (campaign_id)

## Benefits

Flexibility: Campaign owners can accept donations in their preferred asset
Multi-token Support: Enables campaigns for different cryptocurrencies/tokens
User Experience: Better alignment with campaign owner preferences

### PR DESCRIPTION
**Pull request by Akshola00**
closes #116 

* The `create_campaign` function now **requires a `donation_token: ContractAddress`** parameter, making each campaign **support its own ERC-20 token**.

#### 1. **`Errors` Module**

*  Added new error:
  Used to validate the provided token address is not zero.

#### 2. **`CampaignDonation` Contract**

*  Updated `create_campaign` function signature:

  ```cairo
  fn create_campaign(
    ref self: ContractState,
    campaign_ref: felt252,
    target_amount: u256,
    donation_token: ContractAddress,
  ) -> u256
  ```

*  Updated the internal `_create_campaign` method to accept and store `donation_token`:

  ```cairo
  fn _create_campaign(
    ref self: ContractState,
    campaign_ref: felt252,
    target_amount: u256,
    donation_token: ContractAddress,
  ) -> u256
  ```

*  Added validation to reject zero address tokens:

* Modified logic in places like `donate_to_campaign`, `withdraw_funds`, etc., to use `campaign.donation_token` instead of a globally stored token:

#### 3. **Interface (`ICampaignDonation`)**

*  Updated interface to match the new `create_campaign` signature with `donation_token`.

#### 4. **Test File (`test_campaign_donation.cairo`)**

*  Updated all calls to `create_campaign(...)` to include the `donation_token` as an extra argument.
*  Added a new test to assert that an invalid donation token (zero address) causes a panic.

*  **Supports multiple tokens** per campaign instead of relying on a single global token.
*  Prevents campaigns with invalid token addresses from being created.
*  Improves test coverage and correctness by updating all existing test cases and adding necessary edge cases.
*  Ensures backward compatibility is broken in a controlled way (tests are updated accordingly).

### CODE REVIEW

**Comment by Akshola00** [comment]
@mubarak23 ptal

File: src/campaign_donation.cairo [hunk 1]
```diff
@@ -434,11 +438,15 @@ pub mod CampaignDonation {
...
-            ref self: ContractState, campaign_ref: felt252, target_amount: u256,
+            ref self: ContractState,
+            campaign_ref: felt252,
+            target_amount: u256,
+            donation_token: ContractAddress,
         ) -> u256 {
             assert(campaign_ref != '', CAMPAIGN_REF_EMPTY);
             assert(!self.campaign_refs.read(campaign_ref), CAMPAIGN_REF_EXISTS);
             assert(target_amount > 0, ZERO_AMOUNT);
+            assert(donation_token.is_non_zero(), INVALID_DONATION_TOKEN);
```
[the remark this answers is not in the document]
Comment (Akshola00) [reply src/campaign_donation.cairo:449]: alright sir
